File: graphql-api/src/main/java/io/prizy/publicapi/graphql/scalars/LongScalar.java
package io.prizy.publicapi.graphql.scalars;

import graphql.language.StringValue;
import graphql.schema.Coercing;
import graphql.schema.CoercingParseLiteralException;
import graphql.schema.CoercingParseValueException;
import graphql.schema.CoercingSerializeException;
import graphql.schema.GraphQLScalarType;
import lombok.experimental.UtilityClass;

/**
 * @author Nidhal Dogga
 * @created 6/24/2022 9:21 PM
 */


@UtilityClass
public class LongScalar {

  public GraphQLScalarType SCALAR = GraphQLScalarType.newScalar()
    .name("Long")
    .description("Long integer")
    .coercing(new InstantCoercing())
    .build();

  private class InstantCoercing implements Coercing<Long, String> {

    @Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
      if (dataFetcherResult instanceof Long aLong) {
        return aLong.toString();
      }
      throw new CoercingSerializeException("Cannot serialize: " + dataFetcherResult);
    }

    @Override
    public Long parseValue(Object input) throws CoercingParseValueException {
      return Long.valueOf(input.toString());
    }

    @Override
    public Long parseLiteral(Object input) throws CoercingParseLiteralException {
      if (input instanceof StringValue stringValue) {
        return Long.valueOf(stringValue.getValue());
      }
      throw new CoercingParseLiteralException("Cannot parse value: " + input);
    }
  }

}

File: graphql-core/src/main/kotlin/io/prizy/graphql/scalars/LongScalar.kt
package io.prizy.graphql.scalars

import graphql.language.StringValue
import graphql.schema.Coercing
import graphql.schema.CoercingParseLiteralException
import graphql.schema.CoercingSerializeException
import graphql.schema.GraphQLScalarType

object LongScalar {

  val graphqlScalarType: GraphQLScalarType = GraphQLScalarType.newScalar()
    .name("Long")
    .description("Long integer")
    .coercing(LongCoercing)
    .build()

  object LongCoercing : Coercing<Long, String> {

    override fun serialize(dataFetcherResult: Any): String {
      when (dataFetcherResult) {
        is Long -> return dataFetcherResult.toString()
        else -> throw CoercingSerializeException("Not a valid Long")
      }
    }

    override fun parseValue(input: Any): Long {
      return input.toString().toLong()
    }

    override fun parseLiteral(input: Any): Long {
      when (input) {
        is StringValue -> return input.value.toLong()
        else -> throw CoercingParseLiteralException("Value is not a valid Long")
      }
    }
  }
}